mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h

```diff
@@ -0,0 +1,157 @@
+//===- Polynomial.h - A storage class for polynomial types --------*- C++-*-==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef INCLUDE_MLIR_DIALECT_POLYNOMIAL_IR_POLYNOMIAL_H_
+#define INCLUDE_MLIR_DIALECT_POLYNOMIAL_IR_POLYNOMIAL_H_
+
+#include <utility>
+
+#include "mlir/Support/LLVM.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/ADT/DenseMapInfo.h"
+#include "llvm/ADT/Hashing.h"
+
+namespace mlir {
+
+class MLIRContext;
+
+namespace polynomial {
+
+// This restricts statically defined polynomials to have at most 64-bit
+// coefficients. This may be relaxed in the future, but it seems unlikely one
+// would want to specify 128-bit polynomials statically in the source code.
+constexpr unsigned apintBitWidth = 64;
+
+namespace detail {
+struct PolynomialStorage;
+} // namespace detail
+
+class Monomial {
+public:
+  Monomial(int64_t coeff, uint64_t expo)
+      : coefficient(apintBitWidth, coeff), exponent(apintBitWidth, expo) {}
+
+  Monomial(APInt coeff, APInt expo)
+      : coefficient(std::move(coeff)), exponent(std::move(expo)) {}
+
+  Monomial() : coefficient(apintBitWidth, 0), exponent(apintBitWidth, 0) {}
+
+  bool operator==(const Monomial &other) const {
+    return other.coefficient == coefficient && other.exponent == exponent;
+  }
+  bool operator!=(const Monomial &other) const {
+    return other.coefficient != coefficient || other.exponent != exponent;
+  }
+
+  /// Monomials are ordered by exponent.
+  bool operator<(const Monomial &other) const {
+    return (exponent.ult(other.exponent));
+  }
+
+  // Prints polynomial to 'os'.
+  void print(raw_ostream &os) const;
+
+  friend ::llvm::hash_code hash_value(Monomial arg);
+
+public:
+  APInt coefficient;
+
+  // Always unsigned
+  APInt exponent;
+};
+
+/// A single-variable polynomial with integer coefficients. Polynomials are
+/// immutable and uniqued.
+///
+/// Eg: x^1024 + x + 1
+///
+/// The symbols used as the polynomial's indeterminate don't matter, so long as
+/// it is used consistently throughout the polynomial.
+class Polynomial {
```

Comment:
I intend to implement folding for polynomial ops, and this separates the class that would implement the underlying arithmetic from the attribute class. I'm not sure it's strictly necessary, and I can combine the two if you think that would make more sense.

Reply:
Ah I can see an argument of efficiency to be able to perform arithmetic without recreating / uniquing an attribute on every operation.
Can you document the difference between this and the attribute here?

Reply:
But isn't the `Polynomial` object also uniqued here? This looks similar to `AffineMap`/`AffineMapAttr`, and I am probably missing what the second level of boxing brings us here? (I understand that `Monomial` may be a context-unique object while not being available as an attribute though.)

Reply:
Ah, yes, Polynomial is uniqued. I think this discrepancy is mostly my lack of familiarity with this part of MLIR.

What I originally intended to do was just separate conceptually a polynomial from an attribute (just a habit of treating different things differently), and I was borrowing from AffineMap in building this.

I would be open to combining them or keeping them separate and moving the MLIR Context and uniquing stuff to the attribute. Which is preferred?

Reply:
IIUC, the distinction between `AffineMap` and `AffineMapAttr` is a historic artifact at this point (originally, `AffineMap` was a fundamental part of `MemRefType` and we didn't have fully parameterizable types, let alone types parameterized by an attribute). Maybe there is some aspect that I'm forgetting...

We could have a separate `Polynomial` class, but it doesn't have to be a context-owned or uniqued object. This would be similar to the classes in the Presburger library, which are also related to mathematical concepts.

Let's also hear what @joker-eph thinks.

Reply:
Pinging @joker-eph  for any further thoughts on this.

Reply:
I am returning to this ancient PR after months of distraction. I will go ahead and remove the unique-ing from Polynomial

Reply:
Fixed in 8d7f5d1c978